shell/platform/embedder/embedder_surface_software.h

```diff
@@ -0,0 +1,51 @@
+// Copyright 2018 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_PLATFORM_EMBEDDER_EMBEDDER_SURFACE_SOFTWARE_H_
+#define FLUTTER_SHELL_PLATFORM_EMBEDDER_EMBEDDER_SURFACE_SOFTWARE_H_
+
+#include "flutter/fml/macros.h"
+#include "flutter/shell/gpu/gpu_surface_software.h"
+#include "flutter/shell/platform/embedder/embedder_surface.h"
+
+namespace shell {
+
+class EmbedderSurfaceSoftware final : public EmbedderSurface,
+                                      public GPUSurfaceSoftwareDelegate {
+ public:
+  struct SoftwareDispatchTable {
+    std::function<bool(const void* allocation, size_t row_bytes, size_t height)>
```

Comment:
Add width and pixel format arguments to this callback

Reply:
I don't want to make the pixel format configurable as that includes a fair bit of complexity at the API level. So all allocations will be N32. Towards that end, with width can be inferred from the row bytes.

Reply:
I don't think the format needs to be configurable.  But if this is a callback supplied by the user, then I'd like to reserve the flexibility to change the pixel format or add padding in the bitmap given to the callback.

If the bitmap will always be N32 with no padding, then document that in embedder.h

Reply:
I had documented the pixel format and also added a not about buffer ownership.

About padding: Though it is unlikely that we will ever need to change the padding the buffer, changing the padding but keeping the pixel format the same will result in a change in the row bytes. So I believe the API is its current state is adequate.